contrib/babelfishpg_tsql/src/hooks.c

```diff
@@ -2822,6 +2833,255 @@ pltsql_detect_numeric_overflow(int weight, int dscale, int first_block, int nume
+				 */
+				res = makeStringInfo();
+				appendStringInfo(res, "%s-%s-%s", field[0], field[1], field[2]);
+				field[0] = NULL;
+				field[0] = res->data;
+				ftype[0] = DTK_DATE;
+				nf = nf - 2;
+				/*
+				 * Since the first 3 fields converted to 1,
+				 * skip the attached fields.
+				 */
+				for(int i = 1; i < nf; i++)
+				{
+					field[i] = field[i+2];
+					ftype[i] = ftype[i+2];
+				}
+			}
+		}
+
+		for(int i = 0; i < nf; i++)
+		{
+			char *temp_field;
+			int len;
+			char* different_date_formats[] = {"/", ".", "-"};
+			temp_field = pstrdup(field[i]);
+			len = strlen(temp_field);
+			switch(ftype[i])
+			{
+				case DTK_NUMBER:
+					switch(len)
+					{
+						case 1:
+						case 2:
+							/*
+							 * At this point the digit number of length 1,2 should be considered as
+							 * time only if there is an [ap]m after the number or else
+							 * should throw an error
+							 */
+							if(i == nf - 1 || ftype[i+1] != DTK_STRING)
+								TIME_IN_ERROR();
+							status = RE_compile_and_execute(cstring_to_text("^([ap]m)$"),
+															(char *) field[i+1], strlen(field[i+1]),
+															REG_ADVANCED, DEFAULT_COLLATION_OID,
+															0, NULL);
+							if(!status)
+								TIME_IN_ERROR();
+							/*
+							 * For example if the input is "1 am"
+							 * will be converted to "1:00:00 am".
+							 */
+							res = makeStringInfo();
+							appendStringInfo(res, "%s%s", field[i], ":00:00");
+							field[i] = NULL;
+							field[i] = res->data;
+							ftype[i] = DTK_TIME;
+							break;
+
+						case 4:
+							/*
+							 * If the numeric input is of length 4, then convert
+							 * to year with default format of YYYY-01-01.
+							 * For example if input is "2000" will be converted
+							 * to "2000-01-01".
+							 */
+							res = makeStringInfo();
+							appendStringInfo(res, "%s%s", field[i], "-01-01");
+							field[i] = NULL;
+							field[i] = res->data;
+							ftype[i] = DTK_DATE;
+							break;
+
+						case 6:
+						case 8:
+							res = makeStringInfo();
+							for(int k = 0; k < len; k++)
+							{
+								appendStringInfo(res, "%c", temp_field[k]);
+								if((len == 6 && ((k + 1) % 2 == 0 && k < 5)) ||
+									(len == 8 && (k == 3 || k == 5)))
+										appendStringInfo(res, "%c", '-');
+							}
+							field[i] = NULL;
+							field[i] = res->data;	
+							ftype[i] = DTK_DATE;
+							break;
+						/*
+						 * If the numeric is of length {3,5,7, >8} then an error should be thrown*/
+						default:
+							TIME_IN_ERROR();
+					}
+					break;
+
+				case DTK_DATE:
+					/*
+					 * If the input is of format `Mon{/.-}yyyy{/.-}dd` or `Mon{/.-}dd{/.-}yyyy
+					 * shouldn't be supported.
+					 * Supported date formats are `Mon yyyy dd`, `Mon dd yyyy`, `mm{-/.}dd{-/.}yyyy`, `yyyy{-/.}mm{-/.}dd`
+					 */
+					for(int k = 0 ; k < 3; k++)
+					{
+						temp_field = strtok(temp_field, different_date_formats[k]);
+						if (pg_strcasecmp(field[i], temp_field) != 0)
```

Comment:
when we reach here, is it guaranteed that input is one of the three valid date format only and nothing else?

Reply:
Yes it is guaranteed that an valid date format is passed or else later code doesn't check whether it is valid date format or not it checks that valid date is given or not.